### PR TITLE
Enable -Werror=undef for core

### PR DESCRIPTION
**Pull request by eregon**
https://bugs.ruby-lang.org/issues/17752

### CODE REVIEW

File: include/ruby/internal/has/builtin.h [hunk 1]
```diff
@@ -46,7 +46,7 @@
 #  * __has_builtin  only  since  GCC  10.   This section  can  be  made  more
 #  * granular. */
 # /* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66970 */
-# define RBIMPL_HAS_BUILTIN(_) (RBIMPL_HAS_BUILTIN_ ## _)
+# define RBIMPL_HAS_BUILTIN(_) defined(RBIMPL_HAS_BUILTIN_ ## _)
```
Comment (eregon) [inline include/ruby/internal/has/builtin.h:49]: @shyouhei Unfortunately this doesn't work and fails some CI jobs, e.g.
https://github.com/ruby/ruby/pull/4428/checks?check_run_id=2467096081
```
/home/runner/work/ruby/ruby/src/include/ruby/internal/assume.h:44:1: error: this use of "defined" may not be portable [-Werror=expansion-to-defined]
   44 | #if RBIMPL_HAS_BUILTIN(__builtin_unreachable)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is there maybe a way to ignore `expansion-to-defined` warnings while inside the RBIMPL_HAS_BUILTIN macro?

Another way might be to `#define RBIMPL_HAS_BUILTIN___builtin_assume 0` below, but that might be a non-trivial list.

Do you have any other idea how to fix this?

Comment (shyouhei) [reply include/ruby/internal/has/builtin.h:49]: > Is there maybe a way to ignore expansion-to-defined warnings while inside the RBIMPL_HAS_BUILTIN macro?

No... Sadly.  GCC cannot have `_Pragma` inside of `#if`.  That's the ultimate reason why we hesitated `-Wundef` before: https://github.com/ruby/ruby/blob/ruby_3_0/include/ruby/internal/token_paste.h#L35-L38

> Another way might be to add `#define RBIMPL_HAS_BUILTIN___builtin_assume 0` below, but there be many of those.

Yes this seems to be the "polite" way.  Note also that there are non-GCC non-Clang compilers that partially support __builtin_something.  Oracle Developer Studio is one of such compiler.  They use the "Take config.h definition when available" section below.  You would also like to consider modifying there.

> Do you have any other idea how to fix this?

- Dropping GCC <= 9 is one way.  Not practically possible today I guess, though.
- Write `#if (!defined(__GNUC__) && RBIMPL_HAS_BUILTIN(...))`  everywhere.

**Comment by eregon** [comment]
CI is green now :) (except one transient unrelated failure)
@shyouhei Could you review, or suggest who I should ask to review?

File: include/ruby/internal/has/builtin.h [hunk 2]
```diff
@@ -89,16 +96,20 @@
...
+# define RBIMPL_HAS_BUILTIN___builtin_rotateleft32      0
+# define RBIMPL_HAS_BUILTIN___builtin_rotateleft64      0
+# define RBIMPL_HAS_BUILTIN___builtin_rotateright32     0
+# define RBIMPL_HAS_BUILTIN___builtin_rotateright64     0
 # define RBIMPL_HAS_BUILTIN___builtin_popcountll        HAVE_BUILTIN___BUILTIN_POPCOUNTLL
 # define RBIMPL_HAS_BUILTIN___builtin_sub_overflow      HAVE_BUILTIN___BUILTIN_SUB_OVERFLOW
-# if defined(UNREACHABLE)
-#  define RBIMPL_HAS_BUILTIN___builtin_unreachable 1
-# endif
+# define RBIMPL_HAS_BUILTIN___builtin_unreachable       defined(UNREACHABLE)
```
Comment (shyouhei) [inline include/ruby/internal/has/builtin.h:112]: Seems this should also be worked around.

**Review by shyouhei — COMMENTED** [review]
Core changes are okay to me but wonder if extension libraries have their own policies.  For instance I guess `ext/nkf` could have its upstream.

File: include/ruby/internal/has/builtin.h (hunk 2 above)
Comment (eregon) [reply include/ruby/internal/has/builtin.h:112]: Because of potential -Wexpansion-to-defined, right, I'll fix it.

**Comment by eregon** [comment]
> Core changes are okay to me but wonder if extension libraries have their own policies. For instance I guess `ext/nkf` could have its upstream.

I think it's OK to have minor changes fixing warnings, and I guess nkf isn't updated from upstream often either.
`-Wundef` found at least one bug in a C extension (https://github.com/ruby/ruby/pull/4428/commits/e1eab941aabbc941746c9edf0cf4c16cf3af0d11) so I think it's worth having it enabled.
Specific extensions which want to ignore -Wundef warnings can do `$warnflags << " -Wno-undef"`.

**Pull request merged**